### PR TITLE
Containerization Design Doc

### PR DESCRIPTION
**Pull request by arvindpruthi**

### CODE REVIEW

File: docs/imageMgmtAPI.rst [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+      "stability_tag": "enum"   // [experimental/stable/unstable]
+      }
+
+Refer to `usage of this API <#use-case-image-rampup>`_ for implementing a Canary.
+
+.. _get-rampup-plan:
+
+Get an existing Image Rampup Plan
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Returns a list of matching rampup plans for a given imageType
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:218]: Returns an active rampup plan for the specified image type if there is one.

File: docs/imageMgmtAPI.rst [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+     GET /imageRampup?imageType=spark_job
+
+.. _update-rampup-plan:
+
+Update an existing Image Rampup Plan
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Update the active rampup plan and rampup details for an image type.
+
+- **Method:** PATCH
+- **Request URL:** /imageRampup/{planName}
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:243]: request URL: /imageRampup/{imageType}

File: docs/imageMgmtAPI.rst [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+- **Request Body:**
+
++-----------------+-------------+---------------------------------------------------------+
+|   Field Name    |     Type    |            Description                                  |
++=================+=============+=========================================================+
+| ``activePlan``  | ``boolean`` | ``True`` to mark the plan active, ``False`` otherwise.  |
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:249]: Include activatePlan and forceActivatePlan as well.

File: docs/imageMgmtAPI.rst [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
++-----------------+-------------+---------------------------------------------------------+
+
+Use-Cases and Workflows
+-----------------------
+
+New Jobtype/Platform Binary
+***************************
+When a jobtype/platform developer wants to publish their binary for the 1st time, they need to
+register their **imageType** with Azkaban clusters. Hence, the API to `create image type <#create-image-type>`_
+should be called with the required details.
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:261]: As mentioned in my earlier comment, create image type can be invoked only by Azkaban ADMIN.

File: docs/imageMgmtAPI.rst [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+Following this, a `version of the image type must be published <#create-image-version>`_ to be used within Azkaban.
+This is an Admin API and the privileges for this API are with the Azkaban team.
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:264]: Apart from Azkaban ADMIN, Image type owner with role as ADMIN or MEMBER should be able to publish a new version of an image type.

File: docs/imageMgmtAPI.rst [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+
+Image owner wants to publish a new version
+******************************************
+When a jobtype/platform developer wants to deploy a new version of their binary, they need to invoke the API
+to `create new image version <#create-image-version>`_. All imageType users with role as: "admin" as well as
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:269]: Member role as well.

File: docs/imageMgmtAPI.rst [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+When an image owner wants to rampup a new version of their imageType, they can create a new Ramp-up plan for their
+imageType. Here is the ramp-up process:
+
+1. Only one ramp up plan can be created and activated at a time. The ramp up plan must have a new version (latest)
+   and one or more previous versions. Ramp up percentage for all the versions in a plan must add up to 100.
+#. If a new ramp up plan is created for an image type, the previous plan must be deactivated.
+#. The image versions used in the ramp-up plan must be already published/registered with Azkaban and the corresponding
+   images are expected to be available in Artifactory.
+#. The ramp-up plan and ramp ups can be created using the `Create ramp-up plan API <#create-rampup-plan>`_.
+#. The caller of the API must be registered with role as: ``admin`` for the imageType.
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:285]: and member role as well.

File: docs/imageMgmtAPI.rst [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+
+**Example:**
+
+.. code-block::
+
+  POST /image_rampup/sparkJob --body
+    {
+      "planName": "SparkJobRampupPlan",
+      "description": "Ramping up the new release for spark job type",
+      "activePlan": "True",
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:301]: activatePlan

File: docs/containerization-design.rst [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+* If tests pass, the Executor Server is put in inactive mode, upon which it stops polling for flows except for
+  those pinned to that particular executor.
+* A new Executor Server is launched with new binaries and is marked as active; thereby, resuming normal function.
+
+Issues with Bare Metal Executor Server Model
+--------------------------------------------
+
+Noisy-neighbor Issues (No resource Isolation)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Azkaban’s Job Type model allows users to inject arbitrary code on the executor. It is easy to write a jobtype
```
Comment (sshardool) [inline docs/containerization-design.rst:65]: Maybe `job` instead of `jobtype` ?

File: docs/containerization-design.rst [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+
+1. Azkaban will follow a **Disposable Container** model. This implies that whenever a flow is to be launched, the
+   **dispatch logic** will launch a fresh Pod and the pod is destroyed at the conclusion of the flow.
+#. Isolation is achieved per flow (Not at job level). Jobs/subflows that are a part of a flow, will run within the
+   confines of the pod launched for orchestrating the flow. Job level isolation was explored and rejected:
+
+   * It is very disruptive given the existing Azkaban architecture. Major portions of executor code will have to be
+     re-written to accommodate job level isolation.
+   * It appears too resource hungry to launch a separate pod per job. Separate container within the same pod is
+     possible. But again, this would have required rewriting major parts of flow - job wiring. This is something
```
Comment (sshardool) [inline docs/containerization-design.rst:127]: and orchestrators such as kubernetes don't allow creation and scheduling of individual containers (only pods)

File: docs/containerization-design.rst [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+     that can be reconsidered in future.
+
+#. The pod will be launched with default compute/memory resources, but override parameters will be available to
+   request more resources for the flow orchestration.
+#. For this design iteration, the web server will stay outside of k8s. This does not preclude the need for the
+   web server to talk to flow pods to fetch logs or send control commands (Such as Cancel). To enable this
+   communication, an Envoy Proxy based Ingress Controller is introduced, which will allow the web server to
+   communicate with Flow Pods. There is no need to set node ports for flow pods.
```
Comment (sshardool) [inline docs/containerization-design.rst:135]: Maybe rephrase last sentence to `There is no need to directly expose flow pods outside the _(Kubernetes)_ cluster`

File: docs/containerization-design.rst [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+   * The dynamic selection process will be utilized to provide canary capability for various Azkaban components.
+   * The design also introduces a few Admin APIs to make the task of image management easier.
+
+Design Details
+**************
+
+Image Management
+----------------
+* Azkaban will rely on docker images to create execution environment for flows. In order to satisfy
+  `key Requirement #3 <#Key-Requirements-for-Containerization>`_, the final container image will be constructed
```
Comment (sshardool) [inline docs/containerization-design.rst:152]: Consider replacing `container image` with `pod` as that's more accurate. I.e  `the final pod will be constructed dynamically...`

File: docs/containerization-design.rst [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+* Azkaban Core forms the base docker image layered on top of a base image of choice: such as RHEL7.
+* Each of the above (Platform or jobtypes) will be packaged as a separate docker image layers on top of a simple
+  base image of choice. It is preferable to keep these images very small to optimize download times. Something
+  like a busybox/alpine image should suffice.
+* Individual images for job-types will allow independent development and release for the job-type developers without
+  any dependency on Azkaban. Here is an example image definition for KPJ (Kafka Push Job):
+
+.. code-block:: docker
+
+  FROM container-image-registry.corp.linkedin.com/rhel7-base-image/rhel7-base-image:0.16.9
```
Comment (sshardool) [inline docs/containerization-design.rst:177]: Perhaps remove any references to internal websites ?

File: docs/containerization-design.rst [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+* Register new versions of already registered image types
+* Create/Update ramp-up plans for registered image versions
+
+For details regarding these API, please refer to :file:`docs/imageMgmtAPI.rst`
+
+
+Dispatch Logic
+--------------
+
+.. image:: docs/figures/dispatch_logic.png
```
Comment (sshardool) [inline docs/containerization-design.rst:207]: The box labeled `QueueProcessor Thread` seems unnecessary and can be merged with the `Poll for queued flows`. Consider removing.

File: docs/containerization-design.rst [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+   to authenticate flow containers. This and MySQL credentials will be pulled from Kubernetes secret.
+
+Ingress Controller
+------------------
+
+1. As mentioned in the [Flow Container Section](#flow-container), we will be utilizing the
+   Ambassador Ingress Controller [2]_ as a reverse proxy.
+
+#. The ingress controller will provide necessary routing between web server and the flow pods running on
+   kubernetes infrastructure. A key aspect of this architecture is that the routes between web server and flow pods
```
Comment (sshardool) [inline docs/containerization-design.rst:305]: Maybe add a sentence here: `Ambassador essentially provides a dynamically configurable reverse proxy that allows routing of request from the the web-server to the correct flow pod`.

File: docs/imageMgmtAPI.rst [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+.. code-block::
+
+  Status: 201 Created
+  Header -> Location: /imageTypes/{id}
+
+/imageVersions
+**************
+This endpoint is used to register or list image versions for a given image type. The API
+assumes that image type is already registered (Create API for /imageTypes was called before) with Azkaban.
+Also the image binary must already be uploaded to a location in artifactory, accessible to Azkaban.
```
Comment (sshardool) [inline docs/imageMgmtAPI.rst:82]: Should we clarify if this requirement is enforced ?
I.e Do we actually check if the image is available on artifactory when this API is invoked ?

**Review by sshardool — COMMENTED** [review]
Thanks for the clear and concise document!
Have few comments/suggestions.

File: docs/containerization-design.rst (hunk 10 above)
Comment (arvindpruthi) [reply docs/containerization-design.rst:127]: Don't think this argument will fly as Airflow schedules 1 pod per job as well.

File: docs/containerization-design.rst (hunk 11 above)
Comment (arvindpruthi) [reply docs/containerization-design.rst:135]: Done

File: docs/containerization-design.rst (hunk 12 above)
Comment (arvindpruthi) [reply docs/containerization-design.rst:152]: Done

File: docs/containerization-design.rst (hunk 13 above)
Comment (arvindpruthi) [reply docs/containerization-design.rst:177]: done.

File: docs/containerization-design.rst (hunk 15 above)
Comment (arvindpruthi) [reply docs/containerization-design.rst:305]: done

File: docs/containerization-design.rst (hunk 14 above)
Comment (arvindpruthi) [reply docs/containerization-design.rst:207]: The box represents the start of the thread. If you still think it is un-necessary, I will remove it.

File: docs/imageMgmtAPI.rst (hunk 16 above)
Comment (arvindpruthi) [reply docs/imageMgmtAPI.rst:82]: done.

File: docs/imageMgmtAPI.rst [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
++-----------------+-------------+-------------------------------------------------+
+|  imageVersion   |   String    |                                                 |
++-----------------+-------------+-------------------------------------------------+
+|  description    |   String    |                                                 |
++-----------------+-------------+-------------------------------------------------+
+|  versionState   |   String    | ``NEW``, ``ACTIVE``, ``UNSTABLE``,              |
+|                 |             | ``DEPRECATED``                                  |
++-----------------+-------------+-------------------------------------------------+
+|  releaseTag     |   String    |                                                 |
++-----------------+-------------+-------------------------------------------------+
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:117]: Example payload for create version:

{
    "imagePath": "full image path in the artifactory",
    "imageVersion": "1.6.1",
    "imageType": "spark",
    "description": "spark new version",
    "versionState": "NEW",
    "releaseTag": "1.5.9"
}

File: docs/imageMgmtAPI.rst [hunk 18]
```diff
@@ -0,0 +1,373 @@
...
+
+- **Method:** PATCH
+- **Request URL:** /imageVersions/{imageType}
+- **Request Body:**
+
++------------------+-------------+-----------------------------------------------------+
+|    Field Name    |     Type    |            Description                              |
++==================+=============+=====================================================+
+| ``versionState`` | ``String``  | ``NEW``, ``ACTIVE``, ``UNSTABLE`` or ``DEPRECATED`` |
++------------------+-------------+-----------------------------------------------------+
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:192]: Example payload for create version:

{
    "imageType": "spark",
    "imagePath": "updated full image path",
    "versionState": "ACTIVE"
}

File: docs/imageMgmtAPI.rst [hunk 19]
```diff
@@ -0,0 +1,373 @@
...
+``new``, ``active``, ``unstable`` or “deprecated” etc. The state will be updated during the process of ramp up.
+
+- **Method:** PATCH
+- **Request URL:** /imageVersions/{imageType}
+- **Request Body:**
+
++------------------+-------------+-----------------------------------------------------+
+|    Field Name    |     Type    |            Description                              |
++==================+=============+=====================================================+
+| ``versionState`` | ``String``  | ``NEW``, ``ACTIVE``, ``UNSTABLE`` or ``DEPRECATED`` |
```
Comment (abhisheknath2011) [inline docs/imageMgmtAPI.rst:191]: Update is specifically for versionState and hence mandatory. Other optional fields are path and description.

**Comment by jakhani** [comment]
As discussed offline, can we please have flow chart and ER diagram for image management APIs?
We can also add status endpoint related changes for image management APIs.

**Review by sshardool — DISMISSED** [review]
Thanks for addressing feedback.

**Comment by jakhani** [comment]
Let's cover about flow parameter for remote debug.
Parameter name: java.enable.debug
It can be used to setup remote debug for k8s pod. https://blog.christianposta.com/kubernetes/java-remote-debug-for-applications-running-in-kubernetes/

**Comment by arvindpruthi** [comment]
> As discussed offline, can we please have flow chart and ER diagram for image management APIs?
> We can also add status endpoint related changes for image management APIs.

**Review by jakhani — APPROVED** [review]
Thank you so much for writing such useful and descriptive doc. It is very useful!
